Fix not all traits respecting multiple instances of 'Production'

abcdefg30 opened the pull request:
~~I'll update `ClonesProducedUnits` in a follow-up PR.~~

> Note for clarity: Currently the last Harkonnen missions crash (and that's probably just one case; #14796 "fixed" this for the upcoming playtest.) We want to sort all those issues with multiple instances of `Produces` out for `Next+1`. (I.e. replace all occurrences of `Trait<Produces>` and `TraitInfo<ProducesInfo>`)

abcdefg30 commented:
We decided to revert the commit introducing the issue in the meantime: #14796

abcdefg30 commented:
Note for clarity: Currently the last Harkonnen missions crash (and that's probably just one case; #14796 "fixed" this for the upcoming playtest.) We want to sort all those issues with multiple instances of `Produces` out for `Next+1`. (I.e. replace all occurrences of `Trait<Produces>` and `TraitInfo<ProducesInfo>`)

abcdefg30 committed:
Revert "Add conyard.corrino"

This reverts commit a0d4a0353065d48b84de1cf50f1744391a1d4ae0.

GraionDilach commented:
Why is 9e2b55a needed?

pchote commented:
It was a test case commit that was not removed before merging.  It makes the build-everything cheat more difficult to use because it shows as an exact duplicate of the Harkonnen conyard in the sidebar.  It doesn’t make sense / offer any real benefit to allow the special palace to be built using cheats in MP, especially since we do not do the same for the other special factions.

pchote commented:
Do you have a lua script to exercise the testcase? The lua integration is the main thing that needs testing here.

abcdefg30 commented:
Updated.

GraionDilach commented on OpenRA.Mods.Common/Traits/Render/ProductionBar.cs:
```diff
@@ -40,34 +41,44 @@ public ProductionBar(Actor self, ProductionBarInfo info)
 		void ITick.Tick(Actor self)
 		{
-			var current = queue.CurrentItem();
+			var q = queue;
```
Why do we need this variable?

abcdefg30 replied:
How else do you suggest I handle the case when `queue == null` without reassigning the `queue` variable? (We could e.g. move the `value` assignment in if statements, I thought this is the cleaner approach though.)

GraionDilach commented on OpenRA.Mods.Common/Traits/Render/ProductionBar.cs:
```diff
@@ -40,34 +41,44 @@ public ProductionBar(Actor self, ProductionBarInfo info)
 		void FindQueue()
 		{
-			var type = info.ProductionType ?? self.Info.TraitInfo<ProductionInfo>().Produces.First();
+			// Use all queues when no specific one is given
```
Related to my other nit: wouldn't it make more sense to just FieldLoader.Require the queue type explicitly?

abcdefg30 replied:
Maybe; that sounds like a good idea to simplify the code here. Will have a look.

abcdefg30 commented:
Updated.

GraionDilach commented on OpenRA.Mods.Common/Traits/Render/ProductionBar.cs:
```diff
@@ -17,75 +17,63 @@
+		Color ISelectionBar.GetColor() { return Info.Color; }
 		bool ISelectionBar.DisplayWhenEmpty { get { return false; } }
-
-		public void OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
```
Why is this removed? This breaks captured ClassicProdQueues I think.

GraionDilach commented on OpenRA.Mods.Common/Traits/Render/ProductionBar.cs:
```diff
@@ -17,75 +17,63 @@
+					.FirstOrDefault(q => Info.ProductionType == q.Info.Type);
 			}
-
-			if (queue == null)
```
You should also keep this check around, incase the modder typo.

pchote replied:
We (should) have lint tests for that!

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/RequireProductionType.cs:
```diff
@@ -0,0 +1,63 @@
+			get
+			{
+				return "The 'ProductionBar' trait now requires the 'ProductionType' to be set.\n" +
+				"The value be automatically set to the first value in 'Produces' of the first 'Production' trait.";
```
"The value be"

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleProductionActorsHotkeyLogic.cs:
```diff
@@ -40,7 +40,7 @@ protected override bool OnHotkeyActivated(KeyInput e)
 			var facilities = world.ActorsHavingTrait<Production>()
 				.Where(a => a.Owner == player && !a.Info.HasTraitInfo<BaseBuildingInfo>())
-				.OrderBy(f => f.Info.TraitInfo<ProductionInfo>().Produces.First())
+				.OrderBy(f => f.TraitsImplementing<Production>().First(t => !t.IsTraitDisabled).Info.Produces.First())
```
This will throw an exception if all traits are disabled

pchote commented:
While this would be good to have, I don't think it is critical so bumping off the milestone.   If you can do the fixup/rebase in time then we should still try and get this in.

reaperrr commented:
I'm not entirely sure what bleed looks like in that regard (since this PR hasn't been rebased in ages), but we'll want https://github.com/OpenRA/OpenRA/pull/14795/commits/97cb4807e347c8311d03e6985e3c505da1cef40e if it's not on bleed yet (iirc it was picked to prep-1801 directly).

pchote commented:
Added back to the milestone because of the D2K mission crashes. But IMO it would be safer to cherry pick the fix from the last prep branch instead.

abcdefg30 committed:
Fix PrimaryBuilding changing the status of and from disabled queues

abcdefg30 commented:
Resolved the requests and fixed primary building.

GraionDilach commented:
> it would be safer to cherry pick the fix from the last prep branch instead.

Agreeing as the solution for the release tbh.

abcdefg30 commented:
Rebased.

pchote commented:
Needs a rebase, and the update rule should move to the new directory.

abcdefg30 commented:
Rebased.

pchote committed:
Pass the original init dict to UnitProducedByOther.

abcdefg30 committed:
Overhaul ProductionBar:

- Is now a conditional trait
- Now respects multiple Production trait instances
- ProductionType is now required

pchote commented:
I rebased this (on top of #15743 to get the worst of the rebase conflicts out of the way at once), reordered a few things to simplify reviewing, and fixed a few bugs that I noticed.

The code looks mostly ok, but needs testing before giving a final review.

pchote committed:
Remove unnecessary caching of ProductionAirdropInfo.

abcdefg30 committed:
Fix ProductionFromMapEdge blocking the base Created call.

abcdefg30 committed:
Improve ClonesProducedUnits logic:

- Supports multiple Production trait instances
- Clones the correct faction variant, if defined

abcdefg30 committed:
Ignore disabled production traits.

abcdefg30 committed:
Add Lua API support for actors with multiple Production traits.

abcdefg30 committed:
Support multiple Production traits in WithProductionOverlay.

pchote reviewed (approved):
Rebased again, split out specific commits for the different changes, and tested as well as I can.

LGTM, but would like a second review to cover my changes here before we merge.

abcdefg30 commented:
Your changes look good to me. 👍 / ✅

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d728a12c8f2fdd422fa9e521bbe00595f8caaad1...6ffc68eebdd01a641bb5272b750de2854eba7a90)
